### PR TITLE
CHANGE: Organize abstract config into sections and change comment format to docstrings

### PR DESCRIPTION
**Pull request by T-K-233**
**Related PRs / Issues**:

**Type of change**:
- [ ] Bug fix
- [ ] New feature
- [x] Other enhancement

**Impact**:
- [ ] RTL change
- [ ] Software change (RISC-V software)
- [ ] Build system change
- [ ] Other

**Contributor Checklist**:
- [x] Did you set `main` as the base branch?
- [x] Is this PR's title suitable for inclusion in the changelog and have you added a `changelog:<topic>` label?
- [ ] Did you state the type-of-change/impact?
- [x] Did you delete any extraneous prints/debugging code?
- [x] Did you mark the PR with a `changelog:` label?
- [ ] (If applicable) Did you update the conda `.conda-lock.yml` file if you updated the conda requirements file?
- [ ] (If applicable) Did you add documentation for the feature?
- [ ] (If applicable) Did you add a test demonstrating the PR?

- [ ] (If applicable) Did you mark the PR as `Please Backport`?

### CODE REVIEW

**Comment by T-K-233** [comment]
@jerryz123 I'm not sure if I put the all fragments into correct sections.

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 1]
```diff
@@ -51,38 +57,68 @@ class AbstractConfig extends Config(
...
-  new freechips.rocketchip.subsystem.WithDontDriveBusClocksFromSBus ++ // leave the bus clocks undriven by sbus
-  new freechips.rocketchip.subsystem.WithCoherentBusTopology ++     // hierarchical buses including sbus/mbus/pbus/fbus/cbus/l2
-  new freechips.rocketchip.subsystem.WithDTS("ucb-bar,chipyard", Nil) ++ // custom device name for DTS
-  new freechips.rocketchip.system.BaseConfig)                       // "base" rocketchip system
+
+  // Peripheral section
+  new chipyard.config.WithUART ++                                   /** add a UART */
+  
+  // Core section
+  new chipyard.config.WithBootROM ++                                /** use default bootrom */
```
Comment (jerryz123) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:75]: I think these should technically be in the peripheral section

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 2]
```diff
@@ -51,38 +57,68 @@ class AbstractConfig extends Config(
...
+  new testchipip.boot.WithCustomBootPin ++                          /** add a custom-boot-pin to support pin-driven boot address */
+  new testchipip.boot.WithBootAddrReg ++                            /** add a boot-addr-reg for configurable boot address */
+  
+  // ==================================
+  //   Set up tiles
+  // ==================================
+  // Debug settings
+  new chipyard.config.WithDebugModuleAbstractDataWords(8) ++        /** increase debug module data capacity */
```
Comment (jerryz123) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:83]: This is technically a periperal as well.

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 3]
```diff
@@ -51,38 +57,68 @@ class AbstractConfig extends Config(
...
+  // TODO: add these fragments
+  // new chipyard.config.WithJTAGDTMKey(idcodeVersion = 2, partNum = 0x000, manufId = 0x489, debugIdleCycles = 5) ++
+  // new freechips.rocketchip.subsystem.WithNBreakpoints(2) ++
+  new freechips.rocketchip.subsystem.WithJtagDTM ++                 /** set the debug module to expose a JTAG port */
```
Comment (jerryz123) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:87]: Peripheral

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 4]
```diff
@@ -51,38 +57,68 @@ class AbstractConfig extends Config(
...
+  
+  // Cache settings
+  new chipyard.config.WithL2TLBs(1024) ++                           /** use L2 TLBs */
```
Comment (jerryz123) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:90]: This affects cores, the L2 TLB is inside the cores.

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 5]
```diff
@@ -51,38 +57,68 @@ class AbstractConfig extends Config(
...
+  new freechips.rocketchip.subsystem.WithInclusiveCache ++          /** use Sifive L2 cache */
+  
+  // Memory settings
+  new freechips.rocketchip.subsystem.WithNMemoryChannels(1) ++      /** Default 1 memory channels */
+  new freechips.rocketchip.subsystem.WithNoMMIOPort ++              /** no top-level MMIO master port (overrides default set in rocketchip) */
+  new freechips.rocketchip.subsystem.WithNoSlavePort ++             /** no top-level MMIO slave port (overrides default set in rocketchip) */
+  new freechips.rocketchip.subsystem.WithCoherentBusTopology ++     /** hierarchical buses including sbus/mbus/pbus/fbus/cbus/l2 */
+
+  // Core Settings
+  new freechips.rocketchip.subsystem.WithNExtTopInterrupts(0) ++    /** no external interrupts */
```
Comment (jerryz123) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:100]: Peripheral

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala (hunk 1 above)
Comment (T-K-233) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:75]: To clarify, this is the "core" section on the memory map (see [here](https://docs.google.com/presentation/d/18qrFsHImYO4OJEpl8oQ_g3m2dc6a2Q0LppnVxTX-19I/edit#slide=id.g1c27f282ad8_0_54)). These devices affect CPU behavior as their sole purpose, therefore is categorized as "core" section in the memory map; as opposed to other IO or compute periphery devices.

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala (hunk 4 above)
Comment (T-K-233) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:90]: Is this more cache-related or core-functionality related? Here all the subsections are under "Tile".

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala (hunk 5 above)
Comment (T-K-233) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:100]: I'm not sure why this is peripheral. It's not on the memory address, right?

**Comment by jerryz123** [comment]
It might be easier to converge if we agree on a outline of the organization, and then modify the Abstract config.

A proposal:

- HarnessBinders
- IOBinders
- MMIO Devices
  - I/O devices (GPIO/UART/PWM/I2C/SPI)
  - Debug/bringup (JTAG/SerialTL/CustomBoot)
  - Interrupts (PLIC/CLINT)
- Tile settings
- Memory system settings
  - On-chip memory (Scratchpad, BootROM)
  - Off-chip memory (AXI4)
  - LLC
  - Bus settings
- Clocking/reset

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala (hunk 1 above)
Comment (jerryz123) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:75]: I see, let's converge on a organization then. See my other comment.

**Comment by T-K-233** [comment]
I think this outline looks good in general. However, i want to put everything that is on the memory map to the same section, basically putting the "On-chip memory", "Off-chip memory", and "MMIO Devices" together.

**Comment by T-K-233** [comment]
and clock/reset section will also include power (if any), right?

**Comment by jerryz123** [comment]
> I think this outline looks good in general. However, i want to put everything that is on the memory map to the same section, basically putting the "On-chip memory", "Off-chip memory", and "MMIO Devices" together.

I feel like this is hard to accomplish, since a bunch of random things will alter the memory map

**Comment by T-K-233** [comment]
Could we do this instead?:


- HarnessBinders
- IOBinders
- External Memory and IO Devices
  - Off-chip memory (AXI4 / SerialTL / SPI Flash)
  - MMIO devices (GPIO / UART / PWM / I2C / SPI)
  - DMA devices
- Debug Settings
  -  JTAG
  -  CustomBoot
- Interrupts (CLINT / PLIC)
- Tile settings
  - Core settings (FPU / MultiplyUnit)
- Memory system settings
  - On-chip memory (Scratchpad, BootROM)
  - Other cache settings
  - LLC
  - Bus settings
- PRC Settings
  - clocking
  - reset
  - power
- Base settings
  - chipname
  - BaseConfig


This way, the config fragment corresponds the blocks in the SoC block diagram.

Sections still subject to change. I think we can try and see if it works, and if not, we can always make changes later.

**Comment by jerryz123** [comment]
That organization looks good to me.

**Comment by jerryz123** [comment]
@T-K-233 I made some minor changes.
I also removed `WithL2TLBs`... this option actually has no effect when applied to the AbstractConfig, as it needs to override tile-settings that are already specified

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 6]
```diff
@@ -51,40 +57,91 @@ class AbstractConfig extends Config(
...
+      width = 32                                                       /** serial-tilelink interface with 32 lanes */
     )
   )) ++
-  new testchipip.soc.WithMbusScratchpad(base = 0x08000000,          // add 64 KiB on-chip scratchpad
+  new freechips.rocketchip.subsystem.WithNMemoryChannels(1) ++         /** Default 1 AXI-4 memory channels */
+  new freechips.rocketchip.subsystem.WithNoMMIOPort ++                 /** no top-level MMIO master port (overrides default set in rocketchip) */
+  new freechips.rocketchip.subsystem.WithNoSlavePort ++                /** no top-level MMIO slave port (overrides default set in rocketchip) */
+
+
+
```
Comment (abejgonzalez) [inline generators/chipyard/src/main/scala/config/AbstractConfig.scala:75]: Nit: Consistent spacing everywhere.

**Review by abejgonzalez — APPROVED** [review]
Overall LGTM

File: generators/chipyard/src/main/scala/config/AbstractConfig.scala [hunk 7]
```diff
@@ -51,40 +57,91 @@ class AbstractConfig extends Config(
...
+  // ================================================
+  //   Set up Interrupts
+  // ================================================
+  // CLINT and PLIC related settings goes here
+  new freechips.rocketchip.subsystem.WithNExtTopInterrupts(0) ++    /** no external interrupts */
+
+  // ================================================
+  //   Set up Tiles
+  // ================================================
+  // tile-local settings goes here
```
[the remark this answers is not in the document]
Comment (T-K-233) [reply generators/chipyard/src/main/scala/config/AbstractConfig.scala:99]: As a placeholder. The idea is that ChipLikeRocket config will also use this template, and configs like WithFPU will go here.

**Comment by jerryz123** [comment]
@T-K-233 can you rebase then address Abe's comments

**Comment by T-K-233** [comment]
> @T-K-233 can you rebase then address Abe's comments

Fixed

**Review by jerryz123 — APPROVED** [review]

**Pull request merged**